File: day_4/anagrams/character_match.py
# Part 1
def is_character_match(string1, string2):
	string1_dict = {}
	for char in string1.replace(" ", "").lower():
		if char in string1_dict:
			string1_dict[char] += 1
		else:
			string1_dict[char] = 1
	for char in string2.replace(" ", "").lower():
		if char not in string1_dict:
			return False
		else:
			if string1_dict[char] > 0:
				string1_dict[char] -= 1
			else:
				return False
	return True

# Part 2
def anagrams_for(word, list_of_words):
	list_to_return = []
	for word2 in list_of_words:
		if is_character_match(word, word2):
			list_to_return.append(word2)
	return list_to_return

File: day_4/anagrams/characterMatch.js
exports.isCharacterMatch = function(string1, string2) {
  string1 = string1.toLowerCase().replace(/[^A-Z0-9]/ig, "")
  string2 = string2.toLowerCase().replace(/[^A-Z0-9]/ig, "")
  if (string1.length !==string2.length) {
    return false
  }
  let string1Letters = {}
  for (let char of string1) {
    if (char in string1Letters) {
      string1Letters[char]++
    } else {
      string1Letters[char] = 1
    }
  } 
  for (let char of string2) {
    if (char in string1Letters === false) {
      return false
    } else if (string1Letters[char] > 0) {
      string1Letters[char]--
    } else {
      return false
    } return true
  }
};

exports.anagramsFor = function(word, listOfWords) {
  let anagram_list = []
  for (let listword of listOfWords){
    if (exports.isCharacterMatch(word, listword)) {
      anagram_list.push(listword)
    }
  } return anagram_list
};